lib/modem_info/modem_info.c

```diff
@@ -0,0 +1,444 @@
+	.data_type = AT_PARAM_TYPE_STRING,
+};
+
+static const struct modem_info_data cellid_data = {
+	.cmd = AT_CMD_CELLID,
+	.param_index = CELLID_PARAM_INDEX,
+	.param_count = CELLID_PARAM_COUNT,
+	.data_type = AT_PARAM_TYPE_STRING,
+};
+
+static const struct modem_info_data ip_data = {
+	.cmd = AT_CMD_PDP_CONTEXT,
+	.param_index = IP_ADDRESS_PARAM_INDEX,
+	.param_count = IP_ADDRESS_PARAM_COUNT,
+	.data_type = AT_PARAM_TYPE_STRING,
+};
+
+static const struct modem_info_data uicc_data = {
+	.cmd = AT_CMD_UICC_STATE,
+	.param_index = UICC_PARAM_INDEX,
+	.param_count = UICC_PARAM_COUNT,
+	.data_type = AT_PARAM_TYPE_NUM_SHORT,
+};
+
+static const struct modem_info_data battery_data = {
+	.cmd = AT_CMD_VBAT,
+	.param_index = VBAT_PARAM_INDEX,
+	.param_count = VBAT_PARAM_COUNT,
+	.data_type = AT_PARAM_TYPE_NUM_SHORT,
+};
+
+static const struct modem_info_data temp_data = {
+	.cmd = AT_CMD_TEMP,
+	.param_index = TEMP_PARAM_INDEX,
+	.param_count = TEMP_PARAM_COUNT,
+	.data_type = AT_PARAM_TYPE_NUM_SHORT,
+};
+
+static const struct modem_info_data fw_data = {
+	.cmd = AT_CMD_FW_VERSION,
+	.param_index = FW_PARAM_INDEX,
+	.param_count = FW_PARAM_COUNT,
+	.data_type = AT_PARAM_TYPE_STRING,
+};
+
+static const struct modem_info_data *const modem_data[] = {
+	[MODEM_INFO_RSRP] = &rsrp_data,
+	[MODEM_INFO_BAND] = &band_data,
+	[MODEM_INFO_MODE] = &mode_data,
+	[MODEM_INFO_OPERATOR] = &operator_data,
+	[MODEM_INFO_CELLID] = &cellid_data,
+	[MODEM_INFO_IP_ADDRESS] = &ip_data,
+	[MODEM_INFO_UICC] = &uicc_data,
+	[MODEM_INFO_BATTERY] = &battery_data,
+	[MODEM_INFO_TEMP] = &temp_data,
+	[MODEM_INFO_FW_VERSION] = &fw_data,
+};
+
+static const char *const modem_data_name[] = {
+	[MODEM_INFO_RSRP] = "RSRP",
+	[MODEM_INFO_BAND] = "BAND",
+	[MODEM_INFO_MODE] = "MODE",
+	[MODEM_INFO_OPERATOR] = "OPERATOR",
+	[MODEM_INFO_CELLID] = "CELLID",
+	[MODEM_INFO_IP_ADDRESS] = "IP ADDRESS",
+	[MODEM_INFO_UICC] = "SIM",
+	[MODEM_INFO_BATTERY] = "BATTERY",
+	[MODEM_INFO_TEMP] = "TEMP",
+	[MODEM_INFO_FW_VERSION] = "FW",
+};
+
+static rsrp_cb_t modem_info_rsrp_cb;
+
+static struct at_param_list m_param_list;
+static struct k_thread socket_thread;
+static K_THREAD_STACK_DEFINE(socket_thread_stack, THREAD_STACK_SIZE);
+static struct k_mutex socket_mutex;
+static int at_socket_fd = INVALID_DESCRIPTOR;
+static struct pollfd fds;
+static int nfds;
+
+static int at_cmd_send(int fd, const char *cmd, size_t size, char *resp_buffer)
+{
+	int len;
+
+	k_mutex_lock(&socket_mutex, K_FOREVER);
+
+	LOG_DBG("send: %s\n", cmd);
+	len = send(fd, cmd, size, 0);
+	LOG_DBG("len: %d\n", len);
+	if (len != size) {
+		LOG_DBG("send: failed\n");
+		return -EIO;
+	}
+
+	if (resp_buffer != NULL) {
+		len = recv(fd, resp_buffer, CONFIG_MODEM_INFO_BUFFER_SIZE, 0);
+
+		if ((len < strlen(success)) ||
+			(memcmp(success,
+				&resp_buffer[len-AT_CMD_SUCCESS_SIZE],
+				strlen(success)) != 0)) {
+			LOG_ERR("recv: %s", resp_buffer);
+			return -EIO;
+		}
+	}
+
+	k_mutex_unlock(&socket_mutex);
```

Comment:
There are multiple occurrences where you return before you unlock the mutex. You should probably store any error code in a variable so that your execution path always passes this point, and then return the variable at the bottom.

Reply:
Fixed.